### PR TITLE
create Lidar2D sensor component that publish LaserScan.msg

### PR DESCRIPTION
**Pull request by zakmat**
* reuse existing Lidar code where possible
- [x] adjust API to be non-destructive (introduce new methods instead of modifying existing ones)
- [x] rebase on top of development
- [x] explicitly use -Inf for too close detections

### CODE REVIEW

File: Gems/ROS2/Code/Source/Lidar/ROS2Lidar2DSensorComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        const auto& lidarSystemList = LidarRegistrarInterface::Get()->GetRegisteredLidarSystems();
+        if (lidarSystemList.empty())
+        {
+            AZ_Warning("ROS2LidarSensorComponent", false, "No LIDAR system for the sensor to use.");
+            return {};
+        }
+        return lidarSystemList.front();
+    }
+
+    void ROS2Lidar2DSensorComponent::FetchLidarImplementationFeatures()
```
Comment (michalpelka) [inline Gems/ROS2/Code/Source/Lidar/ROS2Lidar2DSensorComponent.cpp:91]: I would recommend considering:
 - using only one implementation (Scene Query)
 - or test against RGL.
 @adamdbrw do you think that supporting RGL would be beneficial for 2D lasers or not here?
 RGL would use visual meshes and can support hundreds of those 2D lidars.

**Review by michalpelka — COMMENTED** [review]
Looks good, some comments here. 
We need at this point agree if RGL support for 2D range finder is something we want.
If the above is false, we should simply remove support for other implementations.

File: Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h [hunk 2]
```diff
@@ -98,12 +98,27 @@ namespace ROS2
         //! @param range Ray range in meters.
         virtual void ConfigureRayRange(float range) = 0;
 
+        //! Configures ray minimum travel distance.
+        //! @param range Ray range in meters.
+        virtual void ConfigureMinimumRayRange(float range)
+        {
+            AZ_Assert(false, "This Lidar Implementation does not support minimum ray range configurations!");
```
[the remark this answers is not in the document]
Comment (zakmat) [reply Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:105]: This is a stub implementation of new API to allow building with other clients. The actual implementation is overridden in `LidarRaycaster` class

File: Gems/ROS2/Code/Source/Lidar/Lidar2DTemplate.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
```
[the remark this answers is not in the document]
Comment (zakmat) [reply Gems/ROS2/Code/Source/Lidar/Lidar2DTemplate.h:7]: Thanks, fixed

File: Gems/ROS2/Code/Source/Lidar/Lidar2DTemplate.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    //! This is meant to capture differences between different 2D Lidars available on the market.
+    struct Lidar2DTemplate
+    {
+    public:
+        AZ_TYPE_INFO(Lidar2DTemplate, "{B9AAD4F8-7D60-4AFC-8CDD-D9D794DE318E}");
+        static void Reflect(AZ::ReflectContext* context);
+
+        enum class Lidar2DModel
+        {
+            Custom2DLidar,
```
[the remark this answers is not in the document]
Comment (zakmat) [reply Gems/ROS2/Code/Source/Lidar/Lidar2DTemplate.h:27]: Great idea! I used following document to get parameter values: https://download.slamtec.com/api/download/rplidar-s1m1-datasheet/1.7?lang=en

File: Gems/ROS2/Code/Source/Lidar/ROS2Lidar2DSensorComponent.cpp (hunk 1 above)
Comment (zakmat) [reply Gems/ROS2/Code/Source/Lidar/ROS2Lidar2DSensorComponent.cpp:99]: We agreed to allow multiple implementations. Tested with RGL and found compilation issue addressed in one of next commits

File: Gems/ROS2/Code/Source/Lidar/Lidar2DTemplate.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        float m_maxHAngle = 0.0f;
+        //! Number of lasers layers (resolution in horizontal direction)
+        //! Resolution in horizontal direction
+        unsigned int m_numberOfIncrements = 1;
+        //! Minimum range of simulated LIDAR
+        float m_minRange = 0.0f;
+        //! Maximum range of simulated LIDAR
+        float m_maxRange = 0.0f;
+    };
+} // namespace ROS2
```
[the remark this answers is not in the document]
Comment (zakmat) [reply Gems/ROS2/Code/Source/Lidar/Lidar2DTemplate.h:46]: newline is actually there - this is the Github presentation layer

**Review by michalpelka — APPROVED** [review]
Check during integration with #260 . Works good.

**Review by adamdbrw — CHANGES_REQUESTED** [review]
Please apply DRY to PerformRaycast functions as well as the Templates

File: Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp [hunk 6]
```diff
@@ -112,23 +106,72 @@ namespace ROS2
...
+            m_sceneHandle = GetPhysicsSceneFromEntityId(m_sceneEntityId);
+        }
+
+        const AZStd::vector<AZ::Vector3> rayDirections =
+            LidarTemplateUtils::RotationsToDirections(m_rayRotations, lidarTransform);
+
+        AzPhysics::SceneQueryRequests requests = prepareRequests(lidarTransform, rayDirections);
+
+        RaycastResult results;
+        if ((flags & RaycastResultFlags::Points) == RaycastResultFlags::Points) {
```
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/Lidar/LidarRaycaster.cpp:134]: apply formatting please

File: Gems/ROS2/Code/Source/Lidar/LidarTemplate.h [hunk 7]
```diff
@@ -31,6 +31,9 @@ namespace ROS2
             Ouster_OS2_64,
             Velodyne_Puck,
             Velodyne_HDL_32E,
+
```
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/Lidar/LidarTemplate.h:34]: extra newline? Perhaps add a comment: // 2D lidars

File: Gems/ROS2/docs/guides/ros2-gem.md [hunk 8]
```diff
@@ -20,6 +20,7 @@ You can browse Doxygen-generated documentation on [Gem's GitHub page](https://ro
   - ROS2GNSSSensorComponent
   - ROS2IMUSensorComponent
   - ROS2LidarSensorComponent
+  - ROS2Lidar2DSensorComponent
```
Comment (adamdbrw) [inline Gems/ROS2/docs/guides/ros2-gem.md:23]: This is now in the documentation (the user guide here is probably going to be removed). Adding a component page to documentation can be done later.

**Review by adamdbrw — APPROVED** [review]
Looks good! Just re-format it please.

File: Gems/ROS2/docs/guides/ros2-gem.md (hunk 8 above)
Comment (zakmat) [reply Gems/ROS2/docs/guides/ros2-gem.md:23]: It is not clear what I could improve here, for now.

File: Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h [hunk 9]
```diff
@@ -98,12 +112,29 @@ namespace ROS2
...
         //! @param lidarTransform Current transform from global to lidar reference frame.
         //! @return Results of the raycast in form of coordinates in 3D space.
         //! The returned vector size can be anything between zero and size of directions. No hits further than distance will be reported.
         virtual AZStd::vector<AZ::Vector3> PerformRaycast(const AZ::Transform& lidarTransform) = 0;
 
+        //! Schedules a raycast that originates from the point described by the lidarTransform.
+        //! @param lidarTransform Current transform from global to lidar reference frame.
+        //! @param flags Used to request different kinds of data returned by raycast query
+        //! @return Results of the raycast in the requested form.
+        virtual RaycastResult PerformRaycastWithFlags(const AZ::Transform& lidarTransform, RaycastResultFlags flags)
```
Comment (alek-kam-robotec-ai) [inline Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:132]: I would suggest replacing this function with void SetRaycastResultFlags(RaycastResultFlags flags). I've started implementing the rgl-gem implementation of the 2D lidar and setting these flags once before raycast calls could both simplify the rgl-gem-side code and improve performance. RGL has an API call for adjusting the raycast result components and modifying it before each raycast could be moderately costly. Additionally this change would simplify the LidarRaycasterBus interface.

Comment (zakmat) [reply Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:132]: Those are some valid points, I like it! Let me do the rework

File: Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h [hunk 10]
```diff
@@ -83,6 +83,20 @@ namespace ROS2
...
+    };
+
+    //! Bitwise operators for RaycastResultFlags
+    AZ_DEFINE_ENUM_BITWISE_OPERATORS(RaycastResultFlags)
+
+    struct RaycastResult
+    {
+        AZStd::vector<AZ::Vector3> m_points;
+        AZStd::vector<float> m_ranges;
+    };
```
Comment (alek-kam-robotec-ai) [inline Gems/ROS2/Code/Include/ROS2/Lidar/LidarRaycasterBus.h:99]: Nit: Missing newline. Otherwise LGTM.

**Review by adamdbrw — APPROVED** [review]

**Review by pawelbudziszewski — APPROVED** [review]

**Pull request merged**